### ISSUE TITLE
Attr-26 (and friends) parsing is wrong.

### ISSUE THREAD

**Issue by alandekok** [FreeRADIUS/freeradius-server#1883]
# Issue type

- [ ] Defect - Crash or memory corruption.
- [ ] Defect - Non compliance with a standards document, or incorrect API usage.
- [X] Defect - Unexpected behaviour (obvious or verified by project member).
- [ ] Feature request.

# Defect/Feature description

Uses of `tmpl_define_unknown_attr()` seem to be wrong.   How that function uses `fr_dict_unknown_add()` is also wrong.

The code converts unknown attributes to known ones (e.g. `Attr-26 = 0x1234` --> `Vendor-Specific = 0x1234`) without considering whether or not the data being assigned to that attribute can be parsed as the given data type.  I've pushed changes to `pairmake_unknown()` which addresses similar issues there.

e.g.`cond_tokenize.c` calls `tmpl_define_unknown_attr()` for both the LHS and RHS of the comparison.  `map_afrom_cp()` does the same.

The result is that having an `update` section as per the unit tests fails:

    make && KEYWORD=vendor_specific_raw build/make/jlibtool --quiet --mode=execute ./build/bin/local/unit_test_module -D share -d src/tests/keywords/ -i build/tests/keywords/vendor_specific_raw.attrs -f build/tests/keywords/vendor_specific_raw.attrs -xx
    ...
    Debug : (0)      update reply {
    ERROR : (0)          ERROR: Implicit cast failed: Invalid cast from octets to vsa.  Can only cast simple data types.

Previous to commit 46221206b9f7fb, it would hit an assertion in `value_box_cast()`

The underlying issue here is that the following is happening:

* `Attr-26` is being parsed
* then converted to `Vendor-Specific`, via `map->lhs->tmpl_da`
* * because attribute 26 exists in the dictionary.
* the RHS is being parsed at run time, and it doesn't match the LHS, so we have a run-time error

Similar issues happen with the `ipaddr.raw` test I just added in commit 0ac574a513d

    KEYWORD=ipaddr.raw build/make/jlibtool --quiet --mode=execute ./build/bin/local/unit_test_module -D share -d src/tests/keywords/ -i build/tests/keywords/ipaddr.raw.attrs -f build/tests/keywords/ipaddr.raw.attrs -xx

What needs to happen instead is:

* parse LHS
* parse RHS
* If LHS is an unknown attribute, AND defined in the dictionary, AND not a structural attribute
* * THEN try to parse the RHS against the LHS dictionary type.
* * IF the parse is successful, convert LHS and RHS to the new types
* otherwise, leave LHS and RHS alone

And then do the same thing for the RHS, if it's an attribute.

I've tried updating `tmpl_define_unknown_attr()` to keep the unknown `da` in `vpt->tmpl_da`, except that the unknown `da` seems to be magically freed somewhere, and it crashes.

`fr_dict_unknown_add()` is probably also wrong.  It shouldn't silently convert unknown attributes to known ones.  Because it doesn't know if the RHS of the expression is parseable.  If not, the template should be left as unknown.

Even when checking for attribute existence, via:

    if (&Attr-26) {

it shouldn't convert unknown to known.  Instead, it should leave the unknown in place.  The check is then NOT for a `Vendor-Specific` attribute, it's for `Attr-26`, i.e. a RADIUS packet which contains a malformed VSA.

v3.0.x doesn't do this any better.  For `ipaddr.raw`, it complains:

    Invalid vendor name in attribute name "0x010203"

which is less useful.

**Comment by alandekok** [comment]
On checking

    f (&Attr-26) {

that seems to be OK, because the code doesn't try to convert `&Attr-26` to `Vendor-Specific`

**Comment by alandekok** [comment]
After some offline discussion with Arran, we have the following conclusions:

* Attr-26 which is defined in the config files is only referenced by name
* therefore `fr_dict_unknown_add()` should add it *only* by name.
* * `fr_dict_unknown_add()` should also look it up by name, and return the previously defined one if it already exists
* the RADIUS decoder gets unknown / malformed attributes by number
* * it calls `fr_dict_unknown_from_fields()`, and `fr_dict_unknown_afrom_fields()`
* * other things call `fr_dict_unknown_from_oid()` and `fr_dict_unknown_afrom_oid()`, `fr_dict_unknown_from_suboid()`, 
* * all of those functions need to be updated to create a name first, e.g. `Attr-26`, and then look up the attribute by name.  If it exists, return that.  Otherwise, create an unknown attribute.
* * all of the callers of those functions will magically start referencing the `Attr-26` defined from the config files, *or* a use-specific "unknown" attribute

**Comment by alandekok** [comment]
On further inspection, `fr_dict_unknown_from_fields()` don't need changing.

If `Attr-1.2.3.4` defines the hierarchy for the attribute, then `fr_dict_unknown_from_fields()`  will never be called, as the dictionary entries by number exist.  If `Attr-1.2.3.4` isn't used in the configuration files anywhere, then we're fine.

**Comment by alandekok** [comment]
The `Attr-4` tests now pass.  But others fail:

    Mismatch at line 523 of src/tests/unit/condition.txt
        got      : &Attr-26.24757.84.9.5.4 == 0x1a99
        expected : &WiMAX-PFDv2-Src-Port == 6809

The `update` and condition code needs to be fixed to:

* look for these attributes
* * we probably need a special flag: raw 
* try to parse the RHS
* * if it parses, convert LHS to the correct name, and RHS to the correct data type
* * otherwise, leave well enough alone

**Comment by alandekok** [comment]
we probably also want to get rid of `TMPL_TYPE_ATTR_UNDEFINED`.  The new code assumes that all attributes are defined when we bootstrap the modules.

**Comment by arr2036** [comment]
Yes that's fine.  The entry in the union needs to be removed too.

**Comment by alandekok** [comment]
I've fixed more tests.  In general, I think we want unknown attributes such as `Attr-26` or `Vendor-4-Attr-1` to require that the RHS always be a hex string, otherwise we can't parse it.

Allowing `Attr-26.11344.1 = 127.0.0.1` is a bad idea.  No one in their right mind uses it, and allowing it makes things more complicated.

There needs to be tests added so that unknown attributes fail if the RHS is not hex.

The `update-xlat` test is failing, because `map.c`  line 1076 has:

    		slen = tmpl_from_attr_str(&exp_lhs, attr, REQUEST_CURRENT, PAIR_LIST_REQUEST, false, false);


which means "don't allow unknown attributes".  And the new code makes `Vendor-11344-Attr-1` to be an unknown / raw attribute.  So there needs to be some more spelunking to figure that out.

The good news is that `map_cast_from_hex()` does a lot of the casting I was looking for.  So that's less work.

But more still needs to be done before the tests pass.

**Comment by alandekok** [comment]
A lot of the tests in src/tests/keywords/unknown are probably wrong...

    &Vendor-11344-Attr-1 == 127.0.0.1

"raw" attributes should only get used with hex RHS.

And we're probably best to get rid of the `Vendor-` prefix.  TBH, I'm inclined to remove it entirely from v4.  It's awkward, and the OID syntax is much better.  That removes a bunch of complexity, and duplicate ways to represent the same thing.  I don't think many people used the `Vendor-` syntax anyways.   Since v3, the server will always print `Attr-xxx` and never `Vendor-`.

Once that's done, we have some rules:

* `Attr-xx = ...`
* * the RHS MUST be hex.  Using data types is forbidden
* once LHS / RHS have been parsed, there's a post-process function which cleans it up
* * this is done right now in a number of places by calling `map_cast_from_hex()`, except the behavior is slightly different in each case, though the code is largely the same

The new "convert / cast" function does:

* If the LHS is a non-octets attribute, and the RHS is non-hex, just parse it, and return any parse error / sucess
* if the LHS is a non-octets attribute, and the RHS is hex, parse the hex && cast it.  return success / fail from `map_cast_from_hex()`.  i.e. if the cast is successful, we're OK.  If not, it's a parse error, OR the value can't be converted.

For octets attributes, we have the following rules

* is !flags.is_raw, it's just hex.  Parse RHS as hex (and allow strings for laziness), and return success / fail from parse.

We now have only raw attributes

* look up original attribute.  If none, it's just a hex attribute.  Parse as normal octets atrribute (see above)

* try to cast RHS from hex to the type of the non-raw LHS attr
* * _right now `map_cast_from_hex()` returns `false` for BOTH parse error, and "cast failed".  That needs to be fixed_
* * if the cast succeeds, convert LHS to the non-raw LHS attr, convert RHS to the correct data type
* * otherwise, leave well enough alone.

This means that this use-case fails, tho:

    update request {
        Attr-1 := 0x414141
    }
    
    if (&Attr-1 == 0x4141) # nope, can't do this
     
    if (User-Name == "aaa") ... # that works

But I'm fine with that...

The alternative is to just say "too bad" in the conf files.  If you put in `Attr-1 = 0x414141`, it's left as that.  And then the ONLY conversion is done in the modules, when reading from external data sources.

I'm not sure we can do that, given that we want the modules to call the map code.  And it's easier to have one code path instead of two.

**Comment by alandekok** [comment]
After some discussion and analysis, all auto-conversion will be removed from the server.  The use-case of updating dictionary is rare.  We can supply an additional tool (if we so care) to do the conversion.

if the admin is dumb enough to put `&Attr-4 == 0x7f000001` into the config, well, it's left as that.  too bad.

**Comment by arr2036** [comment]
One continuing issue is that raw and unknown attributes can't be distinguished from each other.

Before the heuristics for checking MPPE-Keys were improved, if an incorrect shared secret was provided to radsniff, the send/recv key attributes were sometimes shown as OIDs.

This behaviour was reported as a NetworkRADIUS customer as a bug in radsniff, and it took multiple repeated explanations and demonstrations to convince the customer than it was just an issue with the secret, and not a bug with radsniff or the decoder.

Identifying known but malformed (and possibly, separately, undecryptable) attributes needs to be addressed.

There are a few ways this could be done.

One idea is to have multiple different prefixes i.e. ``Attr-``, ``Raw-``.

Another is to extend the selector syntax ``[ ]`` to allow multiple selectors to be applied in a similar way to xpath.

You can then specify the variant of the dictionary attribute using the selector e.g. ``User-Name[raw][0]``.

Alternative suggestions appreciated...

**Comment by arr2036** [comment]
Properties should probably be prefixed with an ``@``, as in xpath i.e. [@raw] or [@malformed].  Looking at this a bit more, it really seems like the ``is_raw`` flag belongs in the VALUE_PAIR itself.

The logic is that a malformed attribute still has the same attribute number and vendor as a well formed attribute, and that's what ``fr_dict_attr_t`` represent.  It's the ``VALUE_PAIR`` itself which is malformed not the fr_dict_attr_t.

The argument against that (from offline discussions with Alan), is that there are many places in the code which expect attributes of a certain type, and just use ``pair_find`` to locate the attribute and access the fields of its value_box_t directly without checking the type (which may not be the expected type in the case of raw attributes).

Using tmpl_expand or tmpl_aexpand removes the danger, as it attempts a cast to the output type, and fails the expansion if the cast fails, but in most cases it would be better if the pair_find functions just didn't return malformed attributes by default.

Seems like this work would be simpler after re-applying the protocol dictionary patches, as it simplifies the VALUE_PAIR lookup API considerably.

**Comment by alandekok** [comment]
On Jan 30, 2017, at 8:15 AM, Arran Cudbard-Bell <notifications@github.com> wrote:
> 
> Properties should probably be prefixed with an @, as in xpath i.e. [@raw] or [@malformed].

  That makes sense.

>  Looking at this a bit more, it really seems like the is_raw flag belongs in the VALUE_PAIR itself.

  Maybe...

> The logic is that a malformed attribute still has the same attribute number and vendor as a well formed attribute, and that's what fr_dict_attr_t represent.

  No.  The dictionary also specifies the data type for the attribute.  See:

https://tools.ietf.org/html/rfc8044

  I *can* claim to be authoritative in this matter. :)

>  It's the VALUE_PAIR itself which is malformed not the fr_dict_attr_t.

  The point is that the value pair is malformed, and is therefore *not* the thing you think it is.  If you receive attribute 4 of length 1, it's *not* NAS-IP-Address.

> The argument against that (from offline discussions with Alan), is that there are many places in the code which expect attributes of a certain type, and just use pair_find to locate the attribute and access the fields of its value_box_t directly without checking the type (which may not be the expected type in the case of raw attributes).
> 
> Using tmpl_expand or tmpl_aexpand removes the danger, as it attempts a cast to the output type, and fails the expansion if the cast fails,

 Only if we replace all uses of VPs with templates.  I don't see that as being helpful.

  And doing dynamic casts is the whole C++ "object oriented flexibility" crap that I hate.

  We *know* that NAS-IP-Address is an IP address.  Modules requiring NAS-IP-Address should have an API that when they ask for NAS-IP-Address, and they either get it (a good one), or they don't get it (bad one, or non-existent).

  Getting crap and then making the module / tmpl code "cast" it to the right thing, and then have the cast fail is just a huge amount of additional work, (CPU, code, debugging) for net negative benefit.  Any time the API is accidentally used wrong (e.g. missing cast), you now have the chance for an external exploit.

  Maybe you mean something else, but it's not clear.

> but in most cases it would be better if the pair_find functions just didn't return malformed attributes by default.

  Which you can trivially do today.  The modules shouldn't be getting attributes by number, they should do the lookups by name when they bootstrap, and then cache the da.  They then look up the attributes by da, and are *guaranteed* to get the right thing.

> Seems like this work would be simpler after re-applying the protocol dictionary patches, as it simplifies the VALUE_PAIR lookup API considerably.

  Probably.

**Comment by arr2036** [comment]
> On 30 Jan 2017, at 13:46, Alan DeKok <notifications@github.com> wrote:
> 
> On Jan 30, 2017, at 8:15 AM, Arran Cudbard-Bell <notifications@github.com> wrote:
> >
> > Properties should probably be prefixed with an @, as in xpath i.e. [@raw] or [@malformed].
> 
> That makes sense.
> 
> > Looking at this a bit more, it really seems like the is_raw flag belongs in the VALUE_PAIR itself.
> 
> Maybe...

> 
> > The logic is that a malformed attribute still has the same attribute number and vendor as a well formed attribute, and that's what fr_dict_attr_t represent.
> 
> No. The dictionary also specifies the data type for the attribute. See:
> 
> https://tools.ietf.org/html/rfc8044
> 
> I *can* claim to be authoritative in this matter. :)

   RADIUS implementations often use "dictionaries" to map attribute
   names to type values and define data types for each attribute.  The
   data types in the dictionaries are defined by each implementation but
   correspond to the "ad hoc" data types used in the specifications.

   In effect, implementations have seen the need for well-defined
   data types and have created them.  It is time for RADIUS
   specifications to follow this practice.

The only mention of dictionaries was the above

Maybe a future RFC could define a standard syntax for dictionaries and specify the behaviour for when the incoming attribute doesn't match the formats specified by RFC8044 :)

> > It's the VALUE_PAIR itself which is malformed not the fr_dict_attr_t.
> 
> The point is that the value pair is malformed, and is therefore *not* the thing you think it is. If you receive attribute 4 of length 1, it's *not* NAS-IP-Address.

Sure, but it's still a base RADIUS attribute with a type field of 4. The dictionary attributes are used any time an attribute needs to be retrieved from the request.  Comparisons based on numbers are removed entirely with the protocol dictionary patches applied.

...and that's a good thing, because numbers on their own lack the protocol and hierarchical context to work properly in a multi-protocol server.

You can't have a module lookup attributes in the request with number 1, and expect them to always be User-Name, it has to be resolved attribute 1 in the RADIUS dictionary, and the result of that resolution is used to lookup attributes in the request.

Doing any comparisons based on numbers is prone to breakage.  Doing checks for DA equivalence is also annoying because you have to check the complete hierarchy, you can't just do da_a->attr == da_b->attr, you need to check every ->parent until you hit the root of the dictionary.  This isn't a shortcoming of the code or design, it's just what you end up having to do when you add a bunch of different protocols with different name/number spaces.

That's why I'm leaning more towards using the same fr_dict_attr_t, and putting the flags in the VP itself, and not having multiple versions of the same fr_dict_attr_t.

> 
> > The argument against that (from offline discussions with Alan), is that there are many places in the code which expect attributes of a certain type, and just use pair_find to locate the attribute and access the fields of its value_box_t directly without checking the type (which may not be the expected type in the case of raw attributes).
> >
> > Using tmpl_expand or tmpl_aexpand removes the danger, as it attempts a cast to the output type, and fails the expansion if the cast fails,
> 
> Only if we replace all uses of VPs with templates. I don't see that as being helpful.

It's not, it'd be a pain.

> And doing dynamic casts is the whole C++ "object oriented flexibility" crap that I hate.

Yeah, applying it everywhere is annoying.

The problem with not applying dynamic casts at all, is you end up exposing implementation details to the user.  It'd be confusing if they couldn't set a 64bit attribute with the value from a 32bit attribute for example.

> We *know* that NAS-IP-Address is an IP address. Modules requiring NAS-IP-Address should have an API that when they ask for NAS-IP-Address, and they either get it (a good one), or they don't get it (bad one, or non-existent).

Agreed. That's the better design.

> Getting crap and then making the module / tmpl code "cast" it to the right thing, and then have the cast fail is just a huge amount of additional work, (CPU, code, debugging) for net negative benefit. Any time the API is accidentally used wrong (e.g. missing cast), you now have the chance for an external exploit.

You can't use it wrong, that's the point of using _Generic.  The function is aware of the C type of the memory it got a pointer to, so you can't trigger a buffer overrun.  It's 'safer' than manually passing in the output type, or accessing the fields of the value_box_t directly.

> Maybe you mean something else, but it's not clear.

It was just enumerating the immediately obvious solutions, I agree adding an argument to the pair_find or cursor API would be better.

> > but in most cases it would be better if the pair_find functions just didn't return malformed attributes by default.
> 
> Which you can trivially do today. The modules shouldn't be getting attributes by number,

Loads of them do, that's why I'm looking at rebasing the dictionary patches, because all those instances are fixed as a side effect.

> they should do the lookups by name when they bootstrap, and then cache the da. They then look up the attributes by da, and are *guaranteed* to get the right thing.

That gets to be annoying.  What's actually done in most places (with the patches applied), is the lookups are done with fr_dict_attr_child_by_num, with the parent being the dictionary root.  Else you have to add initialisation functions everywhere.  But yes the result is the same, the lookup is done with the da.

> > Seems like this work would be simpler after re-applying the protocol dictionary patches, as it simplifies the VALUE_PAIR lookup API considerably.
> 
> Probably.

-Arran

**Comment by alandekok** [comment]
On Jan 30, 2017, at 9:59 AM, Arran Cudbard-Bell <notifications@github.com> wrote:
> > The point is that the value pair is malformed, and is therefore *not* the thing you think it is. If you receive attribute 4 of length 1, it's *not* NAS-IP-Address.
> 
> Sure, but it's still a base RADIUS attribute with a type field of 4.

  So it should point to a 'da' of type 4.  But not to NAS-IP-Address...

  If it points to NAS-IP-Address, then for malformed attributes... what do you print?  You'll need to muck with the printing routines to look for "is_raw" flag, and then print a *different name* if that's set.  Which violates all kinds of layers.

  And you have to muck with all of the comparison routines, to deal with comparisons.  "NAS-IP-Address == 127.0.0.1" has to be able to deal with LHS being of type "octets" of length != 4, which will *never* match the RHS.  But it will still go through the steps of finding the malformed NAS-IP-Address, and doing extra work.

  If instead you create an "unknown da", then the print routines Just Work with no mangling. The comparisons Just Work.

> Doing any comparisons based on numbers is prone to breakage. Doing checks for DA equivalence is also annoying because you have to check the complete hierarchy,

  If you can do fr_dict_attr_child_by_num(), you can do fr_dict_attr_child_by_da()

> That's why I'm leaning more towards using the same fr_dict_attr_t, and putting the flags in the VP itself, and not having multiple versions of the same fr_dict_attr_t.

  I think that breaks all kinds of APIs and assumptions.

> The problem with not applying dynamic casts at all, is you end up exposing implementation details to the user. It'd be confusing if they couldn't set a 64bit attribute with the value from a 32bit attribute for example.

  If the user does that cast, it's one thing.  If an ipaddr attribute magically is *actually* one octet long, that's wrong.

> You can't use it wrong, that's the point of using _Generic. The function is aware of the C type of the memory it got a pointer to, so you can't trigger a buffer overrun. It's 'safer' than manually passing in the output type, or accessing the fields of the value_box_t directly.

  For expansions, that works.  What about modules which have RFC mandated behaviour on attributes?  They will never do any dynamic expansion on the *packet data*, because the module is implementing the RFC on the packet data.  The attributes are either correct and handled correctly, or the attribute is wrong, and the module cannot implement the RFC behaviour.

> That gets to be annoying. What's actually done in most places (with the patches applied), is the lookups are done with fr_dict_attr_child_by_num, with the parent being the dictionary root. Else you have to add initialisation functions everywhere. But yes the result is the same, the lookup is done with the da.

  I'd prefer to move to using the da everywhere in the modules.  And, making malformed attributes use a different da.  That way there's no possibility of mis-use.  And the issues raised above don't come in, either.

**Comment by arr2036** [comment]
As we discussed offline, the issue with having multiple variants of attributes is that the names would conflict unless we used a special prefix as we do for unknowns.

But as we've also discussed magic prefixes are ugly from a design perspective.

One other idea that occurred to me, was to re-work combo functionality to register type variants for malformed attributes referenced by the config (or we could add another separate hash table for the malformed attributes).

That means there'd be separate DAs, which can be looked up by name, and then resolved to the PW_TYPE_OCTETS variant (or malformed variant), if the additional selector was provided in the template i.e. ``[@malformed]``.  

The VP printing functions would then print DAs with the is_raw flag as <name>[@malformed], which would be new functional, except we already decorate some VPs with tags when we print them.

--

The plan is to use DAs in modules, but there are some places where that's not possible in which case the global dictionary roots can be exposed, or we can create global structures for the attributes... Though that gets a bit weird with things like MPPE-Keys.

``fr_dict_attr_child_by_da`` just tells you if a da has been already added beneath a parent, it can't be used for lookups, and can probably be removed as it's no longer used anywhere.  It's not interchangeable with ``fr_dict_child_by_num`` which is the only lookup by number function which should remain.
